tfjs-backend-webgpu/src/matmul_splitK_webgpu.ts

```diff
@@ -0,0 +1,304 @@
+    } else {
+      sampleA =
+          `if(coordsInBounds2D(vec2<i32>(row, col), vec2<i32>(uniforms.dimAOuter, uniforms.dimInner))) {
+            value = A[batch* batchASize + col * uniforms.dimAOuter + row];
+          }
+          return value;`;
+    }
+
+    let sampleB;
+    if (this.transposeB === false) {
+      sampleB =
+          `if(coordsInBounds2D(vec2<i32>(row, col), vec2<i32>(uniforms.dimInner, uniforms.dimBOuter))) {
+            value = B[batch * batchBSize + row * uniforms.dimBOuter + col];
+          }
+          return value;`;
+    } else {
+      sampleB =
+          `if(coordsInBounds2D(vec2<i32>(row, col), vec2<i32>(uniforms.dimInner, uniforms.dimBOuter))) {
+            value = B[batch * batchBSize + col * uniforms.dimInner + row];
+          }
+          return value;`;
+    }
+
+    // atomicAdd only supports uint/int type. For float, we use
+    // atomicCompareExchangeWeak to simulate.
+    const atomicAddSnippet = `
+     var oldValue = atomicLoad(&(result[flatIndex]));
+     var exchanged = false;
+     for (; !exchanged;) {
+       let newValueF32 = bitcast<f32>(oldValue) + value;
+       let newValue = bitcast<i32>(newValueF32);
+       let res = atomicCompareExchangeWeak(&(result[flatIndex]), oldValue, newValue);
+       oldValue = res.old_value;
+       exchanged = res.exchanged;
+     }
+     `;
+
+    const userCode = `
+      fn mm_readA(batch: i32, row : i32, col : i32) -> f32 {
+        let batchASize = uniforms.aShape[1] * uniforms.aShape[2];
+        var value = 0.0;
+        ${sampleA}
+      }
+
+      fn mm_readB(batch: i32, row : i32, col : i32) -> f32 {
+        let batchBSize = uniforms.bShape[1] * uniforms.bShape[2];
+        var value = 0.0;
+        ${sampleB}
+      }
+
+      fn mm_write(batch: i32, row : i32, col : i32, valueIn : f32) {
+        if (row < uniforms.dimAOuter && col < uniforms.dimBOuter) {
+          let outCoord = vec3<i32>(batch, row, col);
+          let flatIndex = getOutputIndexFromCoords(outCoord);
+          var value = valueIn;
+          // The problem is that we should initialize output to zero before using.
+          // Otherwise, the original value will be added to the result.
+          ${atomicAddSnippet}
+        }
+      }
+
+      ${this.makeMatMulSplitKSource()}
+    `;
+    return userCode;
+  }
+
+  makeMatMulSplitKSource(): string {
+    const tileAOuter = this.workGroupSize[1] * this.elementsPerThread[1];
+    const tileBOuter = this.workGroupSize[0] * this.elementsPerThread[0];
+    const rowPerThread = this.elementsPerThread[1];
+    const colPerThread = this.elementsPerThread[0];
+    const colPerThreadA = this.tileInner / this.workGroupSize[0];
+    const rowPerThreadB = this.tileInner / this.workGroupSize[1];
+    util.assert(
+        this.tileInner % this.workGroupSize[0] === 0 &&
+            this.tileInner % this.workGroupSize[1] === 0,
+        () =>
+            `tileInner ${this.tileInner} must be divisible by workGroupSize[0]${
```

Comment:
Is it better to show `tileInner` must larger or equal than workgroupsize[0]|[1] directly?